### ISSUE TITLE
Migrate CI Away from travis-ci.org

### ISSUE THREAD

**Issue by dlstadther** [spotify/luigi#3016]
# tl;dr;
The Luigi project needs to migrate off of Travis-CI.org before the end of the year. Where should we go? Travis-CI.com? Github Actions? Other suggestions?

---
# Background
Currently (as of 2020-11-06), Luigi uses travis-ci.org for builds. They have announced transition to travis-ci.com awhile back and recently set a read-only date for travis-ci.org for December 31 2020 [here](https://docs.travis-ci.com/user/migrate/open-source-repository-migration/#frequently-asked-questions).

Additionally, they have altered their billing model (of which i haven't read all the details yet) [here](https://blog.travis-ci.com/2020-11-02-travis-ci-new-billing).

The Luigi project will need to migrate their CI away from travis-ci.org before the end of the year. I'm happy to help with this process, but would like to get feedback (especially from @honnix and other Spotify folks) regarding _where_ to integrate.

My immediate thought is Github Actions given that is a thing now and has seemed to be reasonable popular.

(For those interested, I found out about this from [this Jeff Gerling blog post](https://www.jeffgeerling.com/blog/2020/travis-cis-new-pricing-plan-threw-wrench-my-open-source-works))

**Comment by lallea** [comment]
Gitlab CI has a unique property that might be interesting for Luigi - it allows BYO local workers/runners to be coordinated by a hosted CI/CD pipeline. For most tests, we would want to use shared hosted runners, but supporting local runners would enable companies to contribute runners that run in-situ tests e.g. for validating cloud service integrations, and have the results integrated in the CI pipeline report.

Gitlab CI can be used with a Github repository. I have had a pleasant experience of Gitlab CI so far, but I haven't used it for OSS.

**Comment by dlstadther** [comment]
Tagging some Spotify folks who have committed / released (sorry if i miss some people)
@honnix @ckiosidis @jesperps @NatashaL @narape @zatine

**Comment by dlstadther** [comment]
Looks like Spotify's [newest open source project](https://github.com/spotify/klio) is using Github Actions. For that reason, i'll also tag @econchick to see if there is Spotify push for their open source projects to use Github Actions, or any option is fair game.

**Comment by econchick** [comment]
hey @dlstadther - it's the maintainers' choice. If you'd like to move to GH actions, go for it. (I, for one, am enjoying GH actions FWIW). As spotify GH owner, feel free to ping me if you need help moving forward w/r/t permissions on GH or travis.

**Comment by honnix** [comment]
@lallea I haven't used Gitlab CI so don't know much about it.

GH actions is a reasonable candidate. A caveat is there is no configuration of machine type, so even we would like to pay for more resource, there is nothing to offer from GH. Not sure how ever Luigi CI is.

CircleCI is another option as some other Spotify repos are using that.

**Comment by lallea** [comment]
How are cloud integrations tested today? Methinks it would be valuable to be able for some company that relies heavily on Luigi and say, GCP services to run GCP integration tests as part of pipeline. :)

CI services also differ in their relation to containers. Many services run builds inside containers, and the support for spinning up new containers from a CI test suite varies. This is relevant for e.g. contrib.kubernetes_test, or if we would like to test e.g. integration with databases without maintaining mocks. 

I have had trouble with e.g. CircleCI and Shippable regarding container support. IIRC, Gitlab Kubernetes executor does not support spinning up new containers, but the Docker machine, Virtualbox, and shell executors do.

Since Luigi sometimes breaks due to changes in external services, I think that we have a good opportunity to choose a tool that does not close the door to automated testing against real services and real databases. I understand, however, that influence is proportional to contribution, and I don't contribute as much as I would like. I have been burnt by many CI services, however, so I'm dumping the information that I have. :-)

**Comment by honnix** [comment]
Sorry for taking so long to come back.

We talked about all possible candidates internally at Spotify, so here are some notes we would like to share.

* There are some implications wrt the new billing model from Travis-Ci.com, so we might just take this opportunity migrating away
* From what we understood none of the maintainers (plus the Spotify team owning Luigi) has good knowledge and experience using  Gitlab CI; the benefit of BYO worker is also available in Github Actions; 
* Github Actions has become increasingly popular and it integrates with the rest of Github features well; it would come handy if we ever need to move/rename the repository
* CircleCI could be a backup plan if Github Actions cannot offer reasonable build time

To summarise, we would like to move to Github Actions with a backup plan using CircleCI. What do others think?

**Comment by dlstadther** [comment]
I think this is a good plan @honnix . Is this CI migration something which Spotify employees want to manage given permission needs for the repo? Or something y'all can just enable and accept PR?

If completed by a non-employee, we could dev the CI based on a fork and contribute back pretty straightforward.

---

> it would come handy if we ever need to move/rename the repository
Has there been discussion, internally, about migrating Spotify OSS to its own org? Or possibly releasing the project to an external party?

**Comment by honnix** [comment]
@dlstadther There are some secrets in Travis but I'm not sure they are still being used. If they're no longer needed I think any maintainer can work on it. I'm not entirely sure whether it will work from a fork but we can try.

**Comment by dlstadther** [comment]
Just as an overdue update.

I've not forgotten about this issue. However, I've not been successful at spending reasonable time on this migration.

For progress, I got a skeleton working in Github Actions which works for flake8 tests, but that's it. There's still plenty more work (and likely more challenging work) to be done for moving over the bulk of our tests. Early results suggest most the difficulty will be related to dependency resolution and environment settings.

**Comment by honnix** [comment]
@dlstadther Sorry for replying late. Is travis-ci.org already dead? I think it should be OK that you open a PR and we ship it to "darkload" and iterate, so other will be able to help. What do you think?

**Comment by ravwojdyla** [comment]
@dlstadther @honnix any updates on this? Maybe before the migration to Github Actions (which might require significant amount of work), we should consider moving to travis-ci.com (which should hopefully be easier), and we can always move to Github Actions later?

**Comment by ravwojdyla** [comment]
ping @dlstadther @honnix

**Comment by honnix** [comment]
@ravwojdyla Thanks for the ping. Another member from Spotify's Data Team will look into this.

**Comment by dlstadther** [comment]
Yeah, sorry. I've been too busy with my day job (which doesn't involve Luigi anymore) to be able to spend time on this migration.

**Comment by ravwojdyla** [comment]
👋 @honnix is there any update on this? is there any specific person from the Spotify's Data Team that's a better contact for this? Btw now that the maintainer is "Spotify's Data Team", what's the point of contact with that group?

**Comment by honnix** [comment]
@ravwojdyla Sorry this is taking longer than expected. I will try to get it resolved asap.

**Comment by ravwojdyla** [comment]
👋  @honnix any update on this? Also now that the maintainer is "Spotify's Data Team", what's the point of contact with that group? Should there be a github group that one could contact?

**Comment by narape** [comment]
@ravwojdyla sorry for the delay. I am Nelson from Spotify. We allocated time during this week for working on migrating away from Travis.

**Comment by ravwojdyla** [comment]
👋  Nelson! Nice to hear from you again, thanks for the update!

**Comment by ckiosidis** [comment]
Hey all and apologies to the rest of the maintainers for the notification spam. I tried to keep it to the minimum.
The last couple of days I helped Nelson move to Github Actions.
We merged today the missing build steps and completely removed Travis CI.
Unfortunately we noticed some small discrepancies between the travisCI code coverage and the Github Actions which seems to be isolated in 3 files:
- luigi/contrib/esindex.py
- luigi/contrib/postgres.py
- luigi/contrib/rdbms.py

We are [merging now the progress](https://github.com/spotify/luigi/pull/3068) to unblock the build pipelines and enable code coverage as a required step and plan to fix this small discrepancy moving on.

I am going to close this ticket. Thank you all for being so patient.

**Comment by dlstadther** [comment]
Thank you @ckiosidis ! Sorry for not being able to spend the time myself to complete this work! Nevertheless, glad to see it done!
